### PR TITLE
crypto: configure key sharing strategy based on DeviceIsolationMode

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes https://github.com/element-hq/element-web/issues/28038
based on the work done on https://github.com/matrix-org/matrix-js-sdk/pull/4407

## Checklist

-   [x] Tests written for new code (and old code if feasible).
-   [ ] New or updated `public`/`exported` symbols have accurate [TSDoc](https://tsdoc.org/) documentation.
-   [x] Linter and other CI checks pass.
-   [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md)).

### CODE REVIEW

**Comment by richvdh** [comment]
Will need to be updated in light of https://github.com/matrix-org/matrix-js-sdk/issues/4428

**Comment by BillCarsonFr** [comment]
> Will need to be updated in light of #4428

Done

File: src/rust-crypto/RoomEncryptor.ts [hunk 1]
```diff
@@ -121,16 +122,20 @@ export class RoomEncryptor {
      * in the room.
      *
      * @param globalBlacklistUnverifiedDevices - When `true`, it will not send encrypted messages to unverified devices
+     * @param deviceIsolationMode - The device isolation mode see {@link DeviceIsolationMode}
      */
-    public async prepareForEncryption(globalBlacklistUnverifiedDevices: boolean): Promise<void> {
+    public async prepareForEncryption(
+        globalBlacklistUnverifiedDevices: boolean,
+        deviceIsolationMode: DeviceIsolationMode = new AllDevicesIsolationMode(false),
```
Comment (richvdh) [inline src/rust-crypto/RoomEncryptor.ts:129]: given that this is only called in one place (except in tests), we should not provide a default value.

Likewise below.

File: src/rust-crypto/RoomEncryptor.ts [hunk 2]
```diff
@@ -121,16 +122,20 @@ export class RoomEncryptor {
      * in the room.
      *
      * @param globalBlacklistUnverifiedDevices - When `true`, it will not send encrypted messages to unverified devices
+     * @param deviceIsolationMode - The device isolation mode see {@link DeviceIsolationMode}
```
Comment (richvdh) [inline src/rust-crypto/RoomEncryptor.ts:125]:
```suggestion
     * @param deviceIsolationMode - The device isolation mode. See {@link DeviceIsolationMode}.
```

Likewise below.

File: src/rust-crypto/RoomEncryptor.ts [hunk 3]
```diff
@@ -174,8 +184,13 @@ export class RoomEncryptor {
      *
      * @param logger - a place to write diagnostics to
      * @param globalBlacklistUnverifiedDevices - When `true`, it will not send encrypted messages to unverified devices
```
Comment (richvdh) [inline src/rust-crypto/RoomEncryptor.ts:186]:
```suggestion
     * @param globalBlacklistUnverifiedDevices - When `true`, and `deviceIsolationMode` is `AllDevicesIsolationMode`, will not send encrypted messages to unverified devices.
           Ignored when  `deviceIsolationMode` is `OnlySignedDevicesIsolationMode`.
```

likewise above.

Comment (richvdh) [inline src/rust-crypto/RoomEncryptor.ts:186]:
```suggestion
     * @param globalBlacklistUnverifiedDevices - When `true`, and `deviceIsolationMode` is `AllDevicesIsolationMode`, will not send encrypted messages to unverified devices.
           Ignored when  `deviceIsolationMode` is `OnlySignedDevicesIsolationMode`.
```

likewise above.

**Review by richvdh — COMMENTED** [review]
looks good to me in general. A couple of nits

File: src/rust-crypto/RoomEncryptor.ts (hunk 1 above)
Comment (BillCarsonFr) [reply src/rust-crypto/RoomEncryptor.ts]: https://github.com/matrix-org/matrix-js-sdk/pull/4425/commits/670fe2e5d84dcc120cc280ee98b1cc0ae024592d

File: src/rust-crypto/RoomEncryptor.ts [hunk 4]
```diff
@@ -120,9 +120,10 @@ export class RoomEncryptor {
      *
      * This ensures that we have a megolm session ready to use and that we have shared its key with all the devices
      * in the room.
-     *
-     * @param globalBlacklistUnverifiedDevices - When `true`, it will not send encrypted messages to unverified devices
-     * @param deviceIsolationMode - The device isolation mode see {@link DeviceIsolationMode}
+     * @param globalBlacklistUnverifiedDevices - When `true`, and `deviceIsolationMode` is `AllDevicesIsolationMode`,
+     * will not send encrypted messages to unverified devices.
+     * Ignored when `deviceIsolationMode` is `OnlySignedDevicesIsolationMode`.
+     * @param deviceIsolationMode - The device isolation mode. See {@link DeviceIsolationMode}
```
Comment (richvdh) [inline src/rust-crypto/RoomEncryptor.ts:126]:
```suggestion
     * @param deviceIsolationMode - The device isolation mode. See {@link DeviceIsolationMode}.
```

**Review by richvdh — APPROVED** [review]

**Comment by richvdh** [comment]
LGTM otherwise

**Pull request merged**